File: priv/javascript/generic_bridge.js
#!/usr/bin/env node

/**
 * Generic JavaScript/Node.js Bridge for Snakepit
 * 
 * A minimal, framework-agnostic bridge that demonstrates the protocol
 * without dependencies on any specific external packages.
 * 
 * This can serve as a template for creating your own JavaScript adapters.
 * 
 * To create a custom adapter:
 * 1. Create a new class that inherits from BaseCommandHandler
 * 2. Override getCommands() to define your command mapping
 * 3. Implement your command handler methods
 * 4. Pass an instance of your handler to ProtocolHandler
 * 
 * Example:
 *     class MyCustomHandler extends BaseCommandHandler {
 *         getCommands() {
 *             return {
 *                 ...super.getCommands(),
 *                 myCommand: this.handleMyCommand.bind(this)
 *             };
 *         }
 *         
 *         handleMyCommand(args) {
 *             return { result: "processed", input: args };
 *         }
 *     }
 *     
 *     const handler = new ProtocolHandler(new MyCustomHandler());
 *     handler.run();
 */

const process = require('process');
const Buffer = require('buffer').Buffer;

/**
 * Abstract base class for command handlers.
 * 
 * This provides a clean interface for creating custom adapters that can
 * be plugged into the ProtocolHandler without modifying the core bridge logic.
 */
class BaseCommandHandler {
    constructor() {
        this.startTime = Date.now();
        this.requestCount = 0;
    }
    
    /**
     * Define the command mapping. Subclasses should override this
     * to add or modify command handlers.
     * 
     * @returns {Object} Map of command names to handler functions
     */
    getCommands() {
        return {
            ping: this.handlePing.bind(this)
        };
    }
    
    /**
     * Get list of supported commands.
     * 
     * @returns {Array} List of command names
     */
    getSupportedCommands() {
        return Object.keys(this.getCommands());
    }
    
    /**
     * Process a command and return the result.
     * 
     * @param {string} command - The command name
     * @param {Object} args - Command arguments
     * @returns {Object} Command result
     * @throws {Error} If command is unknown
     */
    processCommand(command, args) {
        this.requestCount++;
        
        const commands = this.getCommands();
        const handler = commands[command];
        
        if (handler) {
            return handler(args);
        } else {
            throw new Error(`Unknown command: ${command}`);
        }
    }
    
    /**
     * Default ping handler that all adapters can use.
     * 
     * @param {Object} args - Command arguments
     * @returns {Object} Ping response
     */
    handlePing(args) {
        return {
            status: "ok",
            pid: process.pid,
            echo: args
        };
    }
}

/**
 * Generic command handler that provides basic commands without external dependencies.
 * This serves as both a working implementation and an example for custom adapters.
 */
class GenericCommandHandler extends BaseCommandHandler {
    constructor() {
        super();
    }
    
    /**
     * Define all commands supported by the generic handler.
     */
    getCommands() {
        return {
            ...super.getCommands(),
            echo: this.handleEcho.bind(this),
            compute: this.handleCompute.bind(this),
            random: this.handleRandom.bind(this),
            info: this.handleInfo.bind(this)
        };
    }
    
    handlePing(args) {
        /**
         * Enhanced ping with additional info.
         */
        return {
            status: "ok",
            bridge_type: "generic_javascript",
            uptime: (Date.now() - this.startTime) / 1000,
            requests_handled: this.requestCount,
            timestamp: Date.now() / 1000,
            node_version: process.version,
            platform: process.platform,
            worker_id: args.worker_id || "unknown",
            echo: args  // Echo back the arguments for testing
        };
    }
    
    handleEcho(args) {
        /**
         * Handle echo command - useful for testing.
         */
        return {
            status: "ok",
            echoed: args,
            timestamp: Date.now() / 1000
        };
    }
    
    handleCompute(args) {
        /**
         * Handle compute command - simple math operations.
         */
        try {
            const operation = args.operation || "add";
            const a = args.a || 0;
            const b = args.b || 0;
            
            let result;
            
            switch (operation) {
                case "add":
                    result = a + b;
                    break;
                case "subtract":
                    result = a - b;
                    break;
                case "multiply":
                    result = a * b;
                    break;
                case "divide":
                    if (b === 0) {
                        throw new Error("Division by zero");
                    }
                    result = a / b;
                    break;
                case "power":
                    result = Math.pow(a, b);
                    break;
                case "sqrt":
                    if (a < 0) {
                        throw new Error("Square root of negative number");
                    }
                    result = Math.sqrt(a);
                    break;
                default:
                    throw new Error(`Unsupported operation: ${operation}`);
            }
            
            // Additional validation
            if (!isFinite(result)) {
                throw new Error(`Invalid result: ${result}`);
            }
            
            return {
                status: "ok",
                operation: operation,
                inputs: { a: a, b: b },
                result: result,
                timestamp: Date.now() / 1000
            };
        } catch (error) {
            throw error;  // Let the protocol handler catch and format the error
        }
    }
    
    handleRandom(args) {
        /**
         * Handle random command - generate random numbers.
         */
        try {
            const type = args.type || "uniform";
            
            let value;
            
            switch (type) {
                case "uniform":
                    const min = args.min || 0;
                    const max = args.max || 1;
                    value = Math.random() * (max - min) + min;
                    break;
                    
                case "integer":
                    const intMin = args.min || 0;
                    const intMax = args.max || 100;
                    value = Math.floor(Math.random() * (intMax - intMin + 1)) + intMin;
                    break;
                    
                case "normal":
                    const mean = args.mean || 0;
                    const std = args.std || 1;
                    // Box-Muller transformation for normal distribution
                    const u1 = Math.random();
                    const u2 = Math.random();
                    const z0 = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
                    value = z0 * std + mean;
                    break;
                    
                default:
                    throw new Error(`Unsupported random type: ${type}`);
            }
            
            return {
                status: "ok",
                type: type,
                parameters: args,
                value: value,
                timestamp: Date.now() / 1000
            };
        } catch (error) {
            throw error;  // Let the protocol handler catch and format the error
        }
    }
    
    handleInfo(args) {
        /**
         * Handle info command - return bridge information.
         */
        return {
            status: "ok",
            bridge_info: {
                name: "Generic Snakepit JavaScript Bridge",
                version: "1.0.0",
                supported_commands: this.getSupportedCommands(),
                uptime: (Date.now() - this.startTime) / 1000,
                total_requests: this.requestCount
            },
            system_info: {
                node_version: process.version,
                platform: process.platform,
                arch: process.arch,
                memory_usage: process.memoryUsage()
            },
            timestamp: Date.now() / 1000
        };
    }
}

class ProtocolHandler {
    /**
     * Handles the wire protocol for communication with Snakepit.
     * 
     * Protocol:
     * - 4-byte big-endian length header
     * - JSON payload
     */
    
    constructor(commandHandler = null) {
        /**
         * Initialize the protocol handler.
         * 
         * @param {BaseCommandHandler} commandHandler - An instance of BaseCommandHandler or its subclasses.
         *                                              If null, uses GenericCommandHandler as default.
         */
        this.commandHandler = commandHandler || new GenericCommandHandler();
        this.stdin = process.stdin;
        this.stdout = process.stdout;
        
        // Set stdin to raw mode for binary reading (only if it's a TTY)
        if (this.stdin.isTTY) {
            this.stdin.setRawMode(true);
        }
        this.readBuffer = Buffer.alloc(0);
        
        // Increase max listeners to handle high concurrency
        this.stdin.setMaxListeners(50);
        this.stdout.setMaxListeners(50);
    }
    
    readMessage() {
        /**
         * Read a message from stdin using the 4-byte length protocol.
         * Returns a Promise that resolves to the parsed message or null.
         */
        return new Promise((resolve) => {
            const tryRead = () => {
                // Do we have at least 4 bytes for the length header?
                if (this.readBuffer.length < 4) {
                    return false;
                }
                
                // Read the length (big-endian 32-bit integer)
                const length = this.readBuffer.readUInt32BE(0);
                
                // Do we have the complete message?
                if (this.readBuffer.length < 4 + length) {
                    return false;
                }
                
                try {
                    // Extract the JSON payload
                    const jsonData = this.readBuffer.subarray(4, 4 + length);
                    const message = JSON.parse(jsonData.toString('utf-8'));
                    
                    // Remove the processed message from buffer
                    this.readBuffer = this.readBuffer.subarray(4 + length);
                    
                    resolve(message);
                    return true;
                } catch (error) {
                    console.error(`Error parsing message: ${error}`, error);
                    resolve(null);
                    return true;
                }
            };
            
            // Try to read with current buffer
            if (tryRead()) {
                return;
            }
            
            // Set up data listener for more input
            const onData = (chunk) => {
                this.readBuffer = Buffer.concat([this.readBuffer, chunk]);
                if (tryRead()) {
                    this.stdin.removeListener('data', onData);
                }
            };
            
            this.stdin.on('data', onData);
            
            // Handle stdin end
            this.stdin.once('end', () => {
                this.stdin.removeListener('data', onData);
                resolve(null);
            });
        });
    }
    
    writeMessage(message) {
        /**
         * Write a message to stdout using the 4-byte length protocol.
         */
        try {
            // Encode JSON
            const jsonData = Buffer.from(JSON.stringify(message), 'utf-8');
            
            // Create length header (big-endian 32-bit)
            const lengthBuffer = Buffer.allocUnsafe(4);
            lengthBuffer.writeUInt32BE(jsonData.length, 0);
            
            // Write length header and JSON payload
            this.stdout.write(lengthBuffer);
            this.stdout.write(jsonData);
            
            return true;
        } catch (error) {
            console.error(`Error writing message: ${error}`);
            return false;
        }
    }
    
    async run() {
        /**
         * Main message loop.
         */
        console.error("Generic JavaScript Bridge started in pool-worker mode");
        
        try {
            while (true) {
                // Read request
                const request = await this.readMessage();
                if (request === null) {
                    break;
                }
                
                // Extract request details
                const requestId = request.id;
                const command = request.command;
                const args = request.args || {};
                
                let response;
                try {
                    // Process command
                    const result = this.commandHandler.processCommand(command, args);
                    
                    // Check if result indicates an error status
                    if (result && result.status === "error") {
                        response = {
                            id: requestId,
                            success: false,
                            error: result.error,
                            timestamp: new Date().toISOString()
                        };
                    } else {
                        // Send success response
                        response = {
                            id: requestId,
                            success: true,
                            result: result,
                            timestamp: new Date().toISOString()
                        };
                    }
                } catch (error) {
                    // Send error response
                    response = {
                        id: requestId,
                        success: false,
                        error: error.message,
                        timestamp: new Date().toISOString()
                    };
                }
                
                // Write response
                if (!this.writeMessage(response)) {
                    break;
                }
            }
        } catch (error) {
            console.error(`Protocol handler error: ${error}`);
            process.exit(1);
        }
    }
}

function main() {
    /**
     * Main entry point.
     */
    if (process.argv.includes("--help")) {
        console.log("Generic Snakepit JavaScript Bridge");
        console.log("Usage: node generic_bridge.js [--mode pool-worker]");
        console.log("");
        console.log("This bridge provides an extensible architecture for creating custom adapters.");
        console.log("See the module docstring for examples on how to create your own adapter.");
        console.log("");
        console.log("Default supported commands:");
        const handler = new GenericCommandHandler();
        handler.getSupportedCommands().forEach(cmd => {
            console.log(`  ${cmd}`);
        });
        return;
    }
    
    // Start protocol handler
    const handler = new ProtocolHandler();
    
    // Handle graceful shutdown
    process.on('SIGINT', () => {
        console.error("Bridge shutting down (SIGINT)");
        process.exit(0);
    });
    
    process.on('SIGTERM', () => {
        console.error("Bridge shutting down (SIGTERM)");
        process.exit(0);
    });
    
    // Handle pipe errors gracefully
    process.on('SIGPIPE', () => {
        console.error("Bridge pipe closed");
        process.exit(0);
    });
    
    // Handle uncaught errors
    process.on('uncaughtException', (error) => {
        console.error(`Bridge error: ${error}`);
        process.exit(1);
    });
    
    // Handle unhandled promise rejections
    process.on('unhandledRejection', (reason, promise) => {
        console.error(`Unhandled rejection at: ${promise}, reason: ${reason}`);
        process.exit(1);
    });
    
    // Start the main loop
    handler.run().catch((error) => {
        console.error(`Bridge error: ${error}`);
        process.exit(1);
    });
}

if (require.main === module) {
    main();
}

File: priv/python/generic_bridge.py
#!/usr/bin/env python3
"""
Generic Python Bridge for Snakepit

A minimal, framework-agnostic bridge that demonstrates the protocol
without dependencies on any specific ML framework like DSPy.

This can serve as a template for creating your own adapters.

To create a custom adapter:
1. Create a new class that inherits from BaseCommandHandler
2. Override _register_commands() to register your command handlers
3. Implement your command handler methods
4. Pass an instance of your handler to ProtocolHandler

Example:
    class MyCustomHandler(BaseCommandHandler):
        def _register_commands(self):
            self.register_command("my_command", self.handle_my_command)
        
        def handle_my_command(self, args):
            return {"result": "processed", "input": args}
    
    handler = ProtocolHandler(MyCustomHandler())
    handler.run()
"""

import sys
import json
import struct
import time
import signal
import select
import os
from datetime import datetime
from typing import Dict, Any, Optional, Callable
from abc import ABC, abstractmethod


class BaseCommandHandler(ABC):
    """
    Abstract base class for command handlers.
    
    This provides a clean interface for creating custom adapters that can
    be plugged into the ProtocolHandler without modifying the core bridge logic.
    """
    
    def __init__(self):
        self.start_time = time.time()
        self.request_count = 0
        self._command_registry = {}
        self._register_commands()
    
    @abstractmethod
    def _register_commands(self):
        """
        Register all supported commands. Subclasses should override this
        to register their command handlers.
        
        Example:
            self.register_command("ping", self.handle_ping)
            self.register_command("compute", self.handle_compute)
        """
        pass
    
    def register_command(self, command: str, handler: Callable[[Dict[str, Any]], Dict[str, Any]]):
        """Register a command handler."""
        self._command_registry[command] = handler
    
    def get_supported_commands(self) -> list:
        """Get list of supported commands."""
        return list(self._command_registry.keys())
    
    def process_command(self, command: str, args: Dict[str, Any]) -> Dict[str, Any]:
        """Process a command and return the result."""
        self.request_count += 1
        
        handler = self._command_registry.get(command)
        if handler:
            return handler(args)
        else:
            return self._handle_unknown_command(command)
    
    def _handle_unknown_command(self, command: str) -> Dict[str, Any]:
        """Handle unknown commands. Can be overridden by subclasses."""
        return {
            "status": "error",
            "error": f"Unknown command: {command}",
            "supported_commands": self.get_supported_commands(),
            "timestamp": time.time()
        }


class GenericCommandHandler(BaseCommandHandler):
    """
    Generic command handler that provides basic commands without external dependencies.
    This serves as both a working implementation and an example for custom adapters.
    """
    
    def _register_commands(self):
        """Register all generic commands."""
        self.register_command("ping", self.handle_ping)
        self.register_command("echo", self.handle_echo)
        self.register_command("compute", self.handle_compute)
        self.register_command("info", self.handle_info)
        
    def handle_ping(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """Handle ping command - basic health check."""
        return {
            "status": "ok",
            "bridge_type": "generic",
            "uptime": time.time() - self.start_time,
            "requests_handled": self.request_count,
            "timestamp": time.time(),
            "python_version": sys.version,
            "worker_id": args.get("worker_id", "unknown"),
            "echo": args  # Echo back the arguments for testing
        }
    
    def handle_echo(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """Handle echo command - useful for testing."""
        return {
            "status": "ok",
            "echoed": args,
            "timestamp": time.time()
        }
    
    def handle_compute(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """Handle compute command - simple math operations."""
        try:
            operation = args.get("operation", "add")
            a = args.get("a", 0)
            b = args.get("b", 0)
            
            if operation == "add":
                result = a + b
            elif operation == "subtract":
                result = a - b
            elif operation == "multiply":
                result = a * b
            elif operation == "divide":
                if b == 0:
                    raise ValueError("Division by zero")
                result = a / b
            else:
                raise ValueError(f"Unsupported operation: {operation}")
            
            return {
                "status": "ok",
                "operation": operation,
                "inputs": {"a": a, "b": b},
                "result": result,
                "timestamp": time.time()
            }
        except Exception as e:
            return {
                "status": "error", 
                "error": str(e),
                "timestamp": time.time()
            }
    
    def handle_info(self, args: Dict[str, Any]) -> Dict[str, Any]:
        """Handle info command - return bridge information."""
        return {
            "status": "ok",
            "bridge_info": {
                "name": "Generic Snakepit Bridge",
                "version": "1.0.0",
                "supported_commands": self.get_supported_commands(),
                "uptime": time.time() - self.start_time,
                "total_requests": self.request_count
            },
            "system_info": {
                "python_version": sys.version,
                "platform": sys.platform
            },
            "timestamp": time.time()
        }


def safe_print(message: str, file=sys.stderr):
    """Safely print a message, ignoring broken pipe errors."""
    try:
        print(message, file=file)
        file.flush()
    except (BrokenPipeError, IOError):
        # Silently ignore broken pipe errors
        pass


class ProtocolHandler:
    """
    Handles the wire protocol for communication with Snakepit.
    
    Protocol:
    - 4-byte big-endian length header
    - JSON payload
    """
    
    def __init__(self, command_handler: Optional[BaseCommandHandler] = None):
        """
        Initialize the protocol handler.
        
        Args:
            command_handler: An instance of BaseCommandHandler or its subclasses.
                           If None, uses GenericCommandHandler as default.
        """
        self.command_handler = command_handler or GenericCommandHandler()
        self.shutdown_requested = False
        # Disable Python's broken pipe error handling
        signal.signal(signal.SIGPIPE, signal.SIG_DFL) if hasattr(signal, 'SIGPIPE') else None
    
    def read_message(self) -> Optional[Dict[str, Any]]:
        """Read a message from stdin using 4-byte length protocol."""
        try:
            
            # Read 4-byte length header
            length_data = sys.stdin.buffer.read(4)
            if len(length_data) != 4:
                # EOF or pipe closed, return empty dict to signal shutdown
                return {}
            
            # Unpack length (big-endian)
            length = struct.unpack('>I', length_data)[0]
            
            # Read JSON payload
            json_data = sys.stdin.buffer.read(length)
            if len(json_data) != length:
                return {}
            
            # Parse JSON
            return json.loads(json_data.decode('utf-8'))
        except (BrokenPipeError, IOError, OSError):
            # Pipe closed, return empty dict to signal shutdown
            return {}
        except Exception as e:
            safe_print(f"Error reading message: {e}")
            return None
    
    def write_message(self, message: Dict[str, Any]) -> bool:
        """Write a message to stdout using the 4-byte length protocol."""
        try:
            # Encode JSON
            json_data = json.dumps(message, separators=(',', ':')).encode('utf-8')
            
            # Write length header (big-endian)
            length = struct.pack('>I', len(json_data))
            sys.stdout.buffer.write(length)
            
            # Write JSON payload
            sys.stdout.buffer.write(json_data)
            sys.stdout.buffer.flush()
            
            return True
        except (BrokenPipeError, IOError, OSError):
            # Pipe closed, exit silently
            return False
        except Exception as e:
            safe_print(f"Error writing message: {e}")
            return False
    
    def request_shutdown(self):
        """Request graceful shutdown of the main loop."""
        self.shutdown_requested = True
    
    def run(self):
        """Main message loop with non-blocking reads."""
        # Only print startup message if stderr is still connected
        if not os.isatty(sys.stderr.fileno()):
            try:
                # Check if we can write to stderr
                sys.stderr.write("")
                sys.stderr.flush()
                safe_print("Generic Bridge started in pool-worker mode")
            except:
                # stderr is closed, skip the message
                pass
        else:
            safe_print("Generic Bridge started in pool-worker mode")
        
        while not self.shutdown_requested:
            # Read request
            request = self.read_message()
            
            if request is None:
                # Error reading message, continue
                continue
            
            if not request:
                # Empty dict signals EOF or pipe closed, exit cleanly
                break
            
            # Extract request details
            request_id = request.get("id")
            command = request.get("command")
            args = request.get("args", {})
            
            try:
                # Process command
                result = self.command_handler.process_command(command, args)
                
                # Send success response
                response = {
                    "id": request_id,
                    "success": True,
                    "result": result,
                    "timestamp": datetime.now().isoformat()
                }
            except Exception as e:
                # Send error response
                response = {
                    "id": request_id,
                    "success": False,
                    "error": str(e),
                    "timestamp": datetime.now().isoformat()
                }
            
            # Write response
            if not self.write_message(response):
                break
        
        # Exit cleanly
        sys.exit(0)


def main():
    """Main entry point."""
    
    # Suppress broken pipe errors globally
    try:
        # Redirect stderr to devnull on shutdown to avoid broken pipe errors
        import atexit
        def suppress_broken_pipe():
            try:
                sys.stderr.close()
            except:
                pass
            try:
                sys.stdout.close()
            except:
                pass
        atexit.register(suppress_broken_pipe)
    except:
        pass
    
    # Create protocol handler
    handler = ProtocolHandler()
    
    # Set up graceful shutdown handler for SIGTERM
    def graceful_shutdown_handler(signum, frame):
        """Handle SIGTERM by requesting shutdown and exiting cleanly."""
        # Request shutdown of the main loop
        handler.request_shutdown()
        # Close streams to prevent broken pipe errors
        try:
            sys.stdout.close()
        except:
            pass
        try:
            sys.stderr.close()
        except:
            pass
        # Exit cleanly
        os._exit(0)
    
    # Register the signal handler for SIGTERM and SIGINT
    signal.signal(signal.SIGTERM, graceful_shutdown_handler)
    signal.signal(signal.SIGINT, graceful_shutdown_handler)
    
    if len(sys.argv) > 1 and sys.argv[1] == "--help":
        print("Generic Snakepit Bridge")
        print("Usage: python generic_bridge.py [--mode pool-worker]")
        print("")
        print("This bridge provides an extensible architecture for creating custom adapters.")
        print("See the module docstring for examples on how to create your own adapter.")
        print("")
        print("Default supported commands:")
        handler = GenericCommandHandler()
        for cmd in handler.get_supported_commands():
            print(f"  {cmd}")
        return
    
    # Start protocol handler
    try:
        handler.run()
    except (KeyboardInterrupt, BrokenPipeError, IOError):
        # Clean shutdown, suppress errors
        os._exit(0)
    except Exception as e:
        safe_print(f"Bridge error: {e}")
        sys.exit(1)


if __name__ == "__main__":
    main()
